File: dotnet/DesignPatterns/BehavioralDesignPatterns/VisitorDesignPattern/VisitorDesignPattern.Example1/DiscountVisitor.cs
namespace VisitorDesignPattern.Example1;

/// <summary>
/// Concrete Visitor 2: Applies discounts based on product type and properties.
/// Electronics: 5% discount (standard)
/// Food: 10% discount for organic products, 3% for others
/// Clothing: 15% discount (seasonal sale)
/// </summary>
public class DiscountVisitor : IVisitor
{
    public decimal TotalDiscount { get; private set; }

    public void Visit(ElectronicsProduct product)
    {
        decimal discount = product.Price * 0.05m;
        TotalDiscount += discount;
        Console.WriteLine($"  [Discount] {product.Name}: 5% discount = -${discount:F2}");
    }

    public void Visit(FoodProduct product)
    {
        decimal discountRate = product.IsOrganic ? 0.10m : 0.03m;
        decimal discount = product.Price * discountRate;
        TotalDiscount += discount;
        string label = product.IsOrganic ? "10% (organic promotion)" : "3%";
        Console.WriteLine($"  [Discount] {product.Name}: {label} discount = -${discount:F2}");
    }

    public void Visit(ClothingProduct product)
    {
        decimal discount = product.Price * 0.15m;
        TotalDiscount += discount;
        Console.WriteLine($"  [Discount] {product.Name} (Size: {product.Size}): 15% seasonal sale = -${discount:F2}");
    }
}

File: nodejs/DesignPatterns/BehavioralDesignPatterns/VisitorDesignPattern/VisitorDesignPattern.Example1/src/DiscountVisitor.ts
import { IVisitor } from './IVisitor';
import { ElectronicsProduct } from './ElectronicsProduct';
import { FoodProduct } from './FoodProduct';
import { ClothingProduct } from './ClothingProduct';

export class DiscountVisitor implements IVisitor {
  private _totalDiscount: number = 0;

  get totalDiscount(): number {
    return this._totalDiscount;
  }

  visitElectronics(product: ElectronicsProduct): void {
    const discount = product.price * 0.05;
    this._totalDiscount += discount;
    console.log(`  [Discount] ${product.name}: 5% discount = -$${discount.toFixed(2)}`);
  }

  visitFood(product: FoodProduct): void {
    const discountRate = product.isOrganic ? 0.10 : 0.03;
    const discount = product.price * discountRate;
    this._totalDiscount += discount;
    const label = product.isOrganic ? '10% (organic promotion)' : '3%';
    console.log(`  [Discount] ${product.name}: ${label} discount = -$${discount.toFixed(2)}`);
  }

  visitClothing(product: ClothingProduct): void {
    const discount = product.price * 0.15;
    this._totalDiscount += discount;
    console.log(`  [Discount] ${product.name} (Size: ${product.size}): 15% seasonal sale = -$${discount.toFixed(2)}`);
  }
}
